### PR TITLE
Throw exception when circular reference is detected in LG imports

### PR DESCRIPTION
**Pull request by Danieladu**
Issue link: https://github.com/microsoft/botbuilder-js/issues/2340

## Description
Having two language generation files that import each other causes an infinite loop.

## Specific Changes
Throw exception when occurs the cycle reference in LG import.
For example: there are two Lg files:
a.lg
```
> Import b.lg
[b](./b.lg)
# a
- a
```
and b.lg
```
> Import a.lg
[a](./a.lg)
# b
- b
```

When parsing the a.lg, using `Templates.ParseFile("a.lg")`, would get a "loop detect exception."

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplatesParser.cs [hunk 1]
```diff
@@ -148,19 +148,22 @@ public static IParseTree AntlrParseTemplates(LGResource resource)
...
         /// <param name="cachedTemplates">Give the file path and templates to avoid parsing and to improve performance.</param>
+        /// <param name="ancestorTemplates">Ancestor visited Templates.</param>
         /// <returns>new <see cref="Templates"/> entity.</returns>
         public static Templates ParseResource(
             LGResource resource,
             ImportResolverDelegate importResolver = null,
             ExpressionParser expressionParser = null,
-            Dictionary<string, Templates> cachedTemplates = null)
+            Dictionary<string, Templates> cachedTemplates = null,
+            Stack<Templates> ancestorTemplates = null)
```
Comment (boydc2014) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplatesParser.cs:158]: parents?

**Comment by boydc2014** [comment]
@feich-ms can you help review this?

**Review by feich-ms — APPROVED** [review]
looks good.

**Review by boydc2014 — APPROVED** [review]

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplatesParser.cs [hunk 2]
```diff
@@ -148,19 +148,22 @@ public static IParseTree AntlrParseTemplates(LGResource resource)
...
         /// <param name="cachedTemplates">Give the file path and templates to avoid parsing and to improve performance.</param>
+        /// <param name="parentTemplates">Parent visited Templates.</param>
         /// <returns>new <see cref="Templates"/> entity.</returns>
         public static Templates ParseResource(
             LGResource resource,
             ImportResolverDelegate importResolver = null,
             ExpressionParser expressionParser = null,
-            Dictionary<string, Templates> cachedTemplates = null)
+            Dictionary<string, Templates> cachedTemplates = null,
+            Stack<Templates> parentTemplates = null)
```
Comment (boydc2014) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplatesParser.cs:158]: Let's keep the public interface unchange, and create an internal method with an extra state: visitedTemplates.

**Review by boydc2014 — CHANGES_REQUESTED** [review]

**Comment by joshgummersall** [comment]
This does not appear directly linked to https://github.com/microsoft/botbuilder-js/issues/2340, would be good to remove the "Fixes" annotation.

**Comment by Danieladu** [comment]
> This does not appear directly linked to [microsoft/botbuilder-js#2340](https://github.com/microsoft/botbuilder-js/issues/2340), would be good to remove the "Fixes" annotation.

Updated the description

**Comment by joshgummersall** [comment]
@Danieladu, we still need a fix for this in JS, correct?

**Comment by Danieladu** [comment]
> @Danieladu, we still need a fix for this in JS, correct?

yes. After this is done, the porting is required.

**Review by boydc2014 — APPROVED** [review]

**Pull request merged**